dev/io.openliberty.microprofile.openapi.internal.common/src/io/openliberty/microprofile/openapi/internal/common/OpenAPIAppConfigProviderImpl.java

```diff
@@ -0,0 +1,189 @@
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ******************************************************************************/
+
+package io.openliberty.microprofile.openapi.internal.common;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.StringUtils;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.kernel.productinfo.ProductInfo;
+
+import io.openliberty.microprofile.openapi.internal.common.services.OpenAPIAppConfigProvider;
+
+@Component(configurationPolicy = ConfigurationPolicy.OPTIONAL, immediate = true, configurationPid = "io.openliberty.microprofile.openapi")
+public class OpenAPIAppConfigProviderImpl implements OpenAPIAppConfigProvider {
+
+    private static final TraceComponent tc = Tr.register(OpenAPIAppConfigProviderImpl.class);
+
+    private static final String INVALID_APP_WARNING = "OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W";
+    private static final String INVALID_MODULE_WARNING = "OPEN_API_SLASH_IN_MODULE_CWWKO1679W";
+
+    private static final String INCLUDE_APP_PROPERTY_NAME = "includeApplication";
+    private static final String EXCLUDE_APP_PROPERTY_NAME = "excludeApplication";
+    private static final String INCLUDE_MODULE_PROPERTY_NAME = "includeModule";
+    private static final String EXCLUDE_MODULE_PROPERTY_NAME = "excludeModule";
+
+    private Optional<String> includedString = null;
+    private Optional<String> excludedString = null;
+
+    private final Set<OpenAPIAppConfigListener> openAPIAppConfigListeners = new TreeSet<OpenAPIAppConfigListener>();
+
+    private boolean issuedBetaMessage;
+
+    @Activate
+    protected void activate(BundleContext context, Map<String, Object> properties) {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(tc, "Initial processing of server.xml");
+            }
+            processModuleConfig(properties);
+        }
+    }
+
+    @Modified
+    protected void modified(Map<String, Object> properties) {
+        if (ProductInfo.getBetaEdition()) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(tc, "Processing update to server.xml");
+            }
+            processModuleConfig(properties);
+            for (OpenAPIAppConfigListener listener : openAPIAppConfigListeners) {
+                listener.processConfigUpdate();
+            }
+        }
+    }
```

Comment:
This class only cares about a subset of properties from the `<mpOpenAPI>` config element. If someone changes a config value that isn't related to module inclusion/exclusion then we shouldn't call the listeners since re-generating all the OpenAPI documents is a relatively expensive operation.